src/renderer/gdi/paint.cpp

```diff
@@ -666,6 +666,64 @@ try
 }
 CATCH_RETURN();
 
+[[nodiscard]] HRESULT GdiEngine::PaintImageSlice(const ImageSlice& imageSlice,
+                                                 const til::CoordType targetRow,
+                                                 const til::CoordType viewportLeft) noexcept
+try
+{
+    LOG_IF_FAILED(_FlushBufferLines());
+    LOG_IF_FAILED(ResetLineTransform());
+
+    const auto& imagePixels = imageSlice.Pixels();
+    if (_imageMask.size() < imagePixels.size())
+    {
+        _imageMask.resize(imagePixels.size());
+    }
+
+    const auto srcCellSize = imageSlice.CellSize();
+    const auto dstCellSize = _GetFontSize();
+    const auto srcWidth = imageSlice.PixelWidth();
+    const auto srcHeight = srcCellSize.height;
+    const auto dstWidth = srcWidth * dstCellSize.width / srcCellSize.width;
+    const auto dstHeight = srcHeight * dstCellSize.height / srcCellSize.height;
+    const auto x = (imageSlice.ColumnOffset() - viewportLeft) * dstCellSize.width;
+    const auto y = targetRow * dstCellSize.height;
+
+    auto bitmapInfo = BITMAPINFO{
+        .bmiHeader = {
+            .biSize = sizeof(BITMAPINFOHEADER),
+            .biWidth = srcWidth,
+            .biHeight = -srcHeight,
+            .biPlanes = 1,
+            .biBitCount = 32,
+            .biCompression = BI_RGB,
+        }
+    };
+
+    auto allOpaque = true;
+    auto allTransparent = true;
+    for (size_t i = 0; i < imagePixels.size(); i++)
+    {
+        const auto opaque = til::at(imagePixels, i).rgbReserved != 0;
+        allOpaque &= opaque;
+        allTransparent &= !opaque;
+        til::at(_imageMask, i) = (opaque ? 0 : 0xFFFFFF);
+    }
+
+    if (allOpaque)
+    {
+        StretchDIBits(_hdcMemoryContext, x, y, dstWidth, dstHeight, 0, 0, srcWidth, srcHeight, imagePixels.data(), &bitmapInfo, DIB_RGB_COLORS, SRCCOPY);
+    }
+    else if (!allTransparent)
+    {
+        StretchDIBits(_hdcMemoryContext, x, y, dstWidth, dstHeight, 0, 0, srcWidth, srcHeight, _imageMask.data(), &bitmapInfo, DIB_RGB_COLORS, SRCAND);
+        StretchDIBits(_hdcMemoryContext, x, y, dstWidth, dstHeight, 0, 0, srcWidth, srcHeight, imagePixels.data(), &bitmapInfo, DIB_RGB_COLORS, SRCPAINT);

```

Comment:
Side note: I wonder if `CreateDIBitmap` followed by `AlphaBlend` or `TransparentBlt` would be faster. We could even cache the `HBITMAP` instance and use `SetDIBits` to update it.

Reply:
It's been a long time since I wrote most of this code, so my recollection is a bit vague, but I'm almost certain I considered both of those options but ended up rejecting them because I couldn't get them to work. Which is not so say that it can't be done - just that _I_ couldn't do it. I'm more than happy for someone else to give it a try if they can make more sense of the documentation than I could.